### ISSUE TITLE
PTB on Serverless Architectures

### ISSUE THREAD

**Issue by JosXa** [python-telegram-bot/python-telegram-bot#1118]
The notion of `serverless computing` means that code is executed in response to incoming webhook triggers without keeping internal application state (stateless).
In order to make ptb work on serverless architectures, the following points need to be considered:

- Allow to serialize/deserialize `chat_data` and `user_data` on _every_ request in `BasePersistence`
- How to provide an interface for job callbacks, usually run by the `job_queue`  scheduler?
- How to keep track of `conversation_handler.conversations`, is this/can this be included in `BasePersistence`?

**Comment by jsmnbom** [comment]
**Allow to serialize/deserialize `chat_data` and `user_data` on _every_ request in `BasePersistence`**
Already possible IIRC


**How to provide an interface for job callbacks, usually run by the `job_queue`  scheduler?**
This is probably where it's gonna be difficult, as jobs currently aren't covered by persistence at all I think.


**How to keep track of `conversation_handler.conversations`, is this/can this be included in `BasePersistence`?**
Already done :)

So in conclusion, the following is needed for optimal and persistent deployment in such environments
- A solution to job persistence
- Other persistence backends than pickle (like a database or redis or similar)

**Comment by JosXa** [comment]
> How to provide an interface for job callbacks, usually run by the job_queue scheduler?

@jsmnbom this was meant more as a way to trigger job callbacks "from the outside", for instance with the platform's own job scheduler as simple lambdas

**Comment by jsmnbom** [comment]
@JosXa Oh, so you want people to just not use the job_queue scheduler, and instead use something external that is persistent?

**Comment by Eldinnie** [comment]
did #1017 solve this?

**Comment by jsmnbom** [comment]
Partly.. Not not quite, since JobQueue is not part of our persistence implementation currently.

**Comment by JosXa** [comment]
Update: Serverless works just fine now. The manual construction of stuff that the `Updater` normally handles for you is a bit cumbersome, but manageable. 
I'd like some prebuilt persistence class that uses something other than files, e.g. with [redis-collections](https://redis-collections.readthedocs.io/en/stable/).

**Comment by cupcakearmy** [comment]
I managed to get persistence done in the job like this

```python
persistence = PicklePersistence(DB_FILE)
updater: Updater = Updater('token', use_context=True, persistence=persistence)

def update_updater_data():
    updater.dispatcher.user_data = persistence.user_data
    updater.dispatcher.update_persistence()
    persistence.flush()

def my_job():
    persistence.user_data......
    update_updater_data()
```
